src/compiler/parse/state/mustache.ts

```diff
@@ -272,8 +276,8 @@ export default function mustache(parser: Parser) {
 
 		parser.allow_whitespace();
 
-		// {#each} blocks must declare a context – {#each list as item}
-		if (type === 'EachBlock') {
+		// {#each} and {#with} blocks must declare a context – {#each list as item}
+		if (type === 'EachBlock' || type === 'WithBlock') {
```

Comment:
in this implementation {#with foo as bar, i (key)} would also work, maybe write separate if (type === 'WithBlock') with comma and brackets excluded?